### PR TITLE
Update SDK to 7.0.100-alpha.1.21480.20

### PR DESCRIPTION
**Pull request by Tratcher**
Weekly SDK update.

### CODE REVIEW

**Review by wtgodbe — APPROVED** [review]

**Comment by wtgodbe** [comment]
@Tratcher you'll probably need all the changes from https://github.com/dotnet/aspnetcore/pull/36783

**Comment by Tratcher** [comment]
Fixes #37151

**Comment by Tratcher** [comment]
```
Running tests: dotnet run --no-restore --project RunTests/RunTests.csproj -- --target Microsoft.AspNetCore.WebSockets.Tests.dll --runtime 7.0.0-ci --queue Ubuntu.1804.Amd64.Open --arch x64 --quarantined false --ef 7.0.0-alpha.1.21501.1 --helixTimeout 00:30:00
It was not possible to find any compatible framework version
The framework 'Microsoft.NETCore.App', version '6.0.0-rc.2.21452.2' (x64) was not found.
  - The following frameworks were found:
      7.0.0-alpha.1.21480.1 at [/datadisks/disk1/work/9AA508EB/p/dotnet-cli/shared/Microsoft.NETCore.App]
      7.0.0-alpha.1.21501.1 at [/datadisks/disk1/work/9AA508EB/p/dotnet-cli/shared/Microsoft.NETCore.App]
```

Helix seems to be stuck on RC2? Any idea where that's specified?

File: eng/helix/content/RunTests/RunTests.csproj
Comment (wtgodbe) [inline eng/helix/content/RunTests/RunTests.csproj:5]: Should use `DefaultNetCoreTargetFramework` instead. This won't work unless the TFM you're using has entries for `net7.0` - if it does work, please update https://github.com/dotnet/aspnetcore/blob/711bdcf96afbbbe122a302ddc641accea0768431/eng/tools/RepoTasks/RepoTasks.tasks#L4 and https://github.com/dotnet/aspnetcore/blob/711bdcf96afbbbe122a302ddc641accea0768431/eng/tools/RepoTasks/RepoTasks.csproj#L4 too

**Comment by Tratcher** [comment]
Linux build failure:
```
Cannot get required symbol SSL_set_alpn_protos from libssl
/mnt/vss/_work/1/s/eng/common/tools.sh: line 475:   139 Aborted                 (core dumped) "$_InitializeBuildTool" "$@"
```

File: eng/helix/content/RunTests/RunTests.csproj
Comment (Tratcher) [reply eng/helix/content/RunTests/RunTests.csproj]: DefaultNetCoreTargetFramework has been used in RepoTasks.csproj before, but not in RepoTasks.cs, but not in RunTests.csproj where it doesn't seem to be defined.
```
/datadisks/disk1/work/A9D20953/p/dotnet-cli/sdk/7.0.100-alpha.1.21480.20/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(92,5): error NETSDK1013: The TargetFramework value '' was not recognized. It may be misspelled. If not, then the TargetFrameworkIdentifier and/or TargetFrameworkVersion properties must be specified explicitly. [/datadisks/disk1/work/A9D20953/w/A34C0943/e/RunTests/RunTests.csproj]
```

Comment (BrennanConroy) [reply eng/helix/content/RunTests/RunTests.csproj:5]: RunTests has a directory.build.props that doesn't consume the main repo one so it doesn't have all the variables that other projects do. Pretty sure there was a good reason for not consuming the main build.props file

Comment (dougbu) [reply eng/helix/content/RunTests/RunTests.csproj:5]: > Pretty sure there was a good reason for not consuming the main build.props file

Yeah, it's intentional to avoid tripping ourselves up when we change `$(DefaultNetCoreTargetFramework)` before we have an SDK that truly supports it. `RepoTasks` and `RunTests` should work w/ the base SDK and not depend on our various workarounds. `net7.0` is fine here.

File: eng/docker/rhel.Dockerfile
Comment (dougbu) [inline eng/docker/rhel.Dockerfile:2]: 👍

**Comment by Tratcher** [comment]
Hmm, the only commas are in the description.
--description="Shared Framework for hosting of Microsoft ASP.NET Core applications. It is open source, cross-platform and is supported by Microsoft. We hope you enjoy using it! If you do, please consider joining the active community of developers that are contributing to the project on GitHub (https://github.com/dotnet/aspnetcore). We happily accept issues and PRs."

We could always try a run-on sentence 😁.

**Comment by dougbu** [comment]
> We could always try a run-on sentence 😁.

Maybe the command supports escaping the comma in some way❔ I'm not sure how to find the Ruby sources but I'm hoping it's more clear when looking at baseparser.rb:673, baseparser.rb:542, or baseparser.rb:515 what it's attempting to parse.

**Comment by Tratcher** [comment]
Hmm
https://github.com/ruby/rexml/blob/v3.2.5/lib/rexml/parsers/baseparser.rb#L515

I'm not sure it's the command, it looks more like an issue loading the ruby code. Does this image have the right version?

**Comment by Tratcher** [comment]
https://github.com/ruby/rexml/issues/69

**Comment by dougbu** [comment]
> Does this image have the right version?

Hmm, `v3.2.5` is the latest tag in ruby/rexml. Probably need to request a new mcr image that either backs up rexml to `v3.2.4` or moves Ruby forward to `2.1`. I don't see a newer rexml on RubyGems.org. Thoughts @janvorli❔

**Comment by Tratcher** [comment]
Per ruby/rexml#69 it looks like ruby 2.5 is required for rexml.

**Comment by TanayParikh** [comment]
> Per [ruby/rexml#69](https://github.com/ruby/rexml/issues/69) it looks like ruby 2.5 is required for rexml.

@Tratcher were you suggesting we need to update the version of ruby used in the CI?

> Hmm, `v3.2.5` is the latest tag in ruby/rexml. Probably need to request a new mcr image that either backs up rexml to `v3.2.4` or moves Ruby forward to `2.1`. I don't see a newer rexml on RubyGems.org. Thoughts @janvorli❔

/ping @janvorli

**Comment by Tratcher** [comment]
We switched from rhel-7-rpmpkg-e1b4a89-20175311035359 to centos-7-rpmpkg-20210714125435-9b5bbc2 for the RPM build step but the new image doesn't have the right version of ruby and needs to be updated/augmented.

**Comment by TanayParikh** [comment]
> Do we need to reach out to someone regarding updating the ruby version in the docker image?

Actually, there's a centos 8 image. Trying that now.

File: eng/docker/rhel.Dockerfile
Comment (TanayParikh) [reply eng/docker/rhel.Dockerfile:2]:
```suggestion
FROM mcr.microsoft.com/dotnet-buildtools/prereqs:centos-8-rpmpkg-20210714125410-daa5116
```

From https://github.com/dotnet/versions/blob/a3164ca1c7f07c6ad990834bd741e16ac24b3ec5/build-info/docker/image-info.dotnet-dotnet-buildtools-prereqs-docker-main.json#L745

**Comment by TanayParikh** [comment]
Linux build passing with the centos 8 image!

image

@dotnet/aspnet-build could I get a final review before merge.

**Comment by Tratcher** [comment]
@dougbu ? I know you'd wanted to avoid the centos-8 image.

**Comment by dougbu** [comment]
> @dougbu ? I know you'd wanted to avoid the centos-8 image.

Yes, I'm concerned about our support matrix when building w/ CentOS 8. I'd much prefer a usable CentOS 7 image.

/cc @janvorli @mthalman @NikolaMilosavljevic

**Comment by janvorli** [comment]
> From my research rhel 7 comes with ruby 2.7+ preinstalled

Hmm, I've just loaded the rhel 7 image mcr.microsoft.com/dotnet-buildtools/prereqs:rhel-7-rpmpkg-e1b4a89-20175311035359 and the `ruby -v` shows `ruby 2.0.0p648 (2015-12-16) [x86_64-linux]`.

I would be surprised if rhel 7 and centos 7 differed in available package versions, I have always believed those two are the same in that respect.

**Comment by janvorli** [comment]
It is possible to install ruby 2.5 on centos 7 though, from the `centos-sclo-rh` repo. You's need to execute anything that uses ruby using `scl enable rh-ruby25` or run the bash shell that you run your stuff in using that command (`scl enable rh-ruby25 bash`) or add `/etc/profile.d/rh-ruby25.sh` script like described here: https://www.server-world.info/en/note?os=CentOS_7&p=ruby25
Would that be feasible?

**Comment by TanayParikh** [comment]
> It is possible to install ruby 2.5 on centos 7 though, from the `centos-sclo-rh` repo. You's need to execute anything that uses ruby using `scl enable rh-ruby25` or run the bash shell that you run your stuff in using that command (`scl enable rh-ruby25 bash`) or add `/etc/profile.d/rh-ruby25.sh` script like described here: https://www.server-world.info/en/note?os=CentOS_7&p=ruby25 Would that be feasible?

I believe that'd require updating the dockerfile [here](https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/20a9d9c027296911bb76ea1f3f3f8eded2978c36/src/centos/7/rpmpkg/Dockerfile#L4-L14) with the ruby25 script / instantiation you suggest. This would likely involve creating and publishing a new image with those changes. Would that be possible? cc/ @dougbu ?

**Comment by dougbu** [comment]
@janvorli 

> As for using CentOS 8, is this image used to build some native components too?

We use the image to build only our two RPM projects: [src/Installers/Rpm/Runtime/Rpm.Runtime.rpmproj](https://github.com/dotnet/aspnetcore/blob/main/src/Installers/Rpm/Runtime/Rpm.Runtime.rpmproj) and [src/Installers/Rpm/TargetingPack/Rpm.TargetingPack.rpmproj](https://github.com/dotnet/aspnetcore/blob/main/src/Installers/Rpm/TargetingPack/Rpm.TargetingPack.rpmproj). Most of the real work is done in [src/Installers/Rpm/Directory.Build.targets](https://github.com/dotnet/aspnetcore/blob/main/src/Installers/Rpm/Directory.Build.targets). From what I can see, `fpm` gets run but not much else. See [a recent log](https://dev.azure.com/dnceng/internal/_build/results?buildId=1406146&view=logs&j=b72e85ab-3386-5aa9-6405-3837662d9688&t=5d75deca-26a1-543d-80c0-3108c555a3cb) for example.

If we can move to CentOS 8 without breaking our RHEL 7 compatibility, I guess I'm :ok: with it. I continue to have nagging doubts about whether `fpm` produces the right RPMs.

> Hmm, I've just loaded the rhel 7 image mcr.microsoft.com/dotnet-buildtools/prereqs:rhel-7-rpmpkg-e1b4a89-20175311035359 and the `ruby -v` shows `ruby 2.0.0p648 (2015-12-16) [x86_64-linux]`.

That's not a big surprise because that image is for RHEL 7.5 and is 3 or 4 years old. We were using it until it stopped working with the OpenSSL expectations in the SDK. That's why we've been messing w/ the image in the first place.

An updated RHEL 7 or CentOS 7 image containing compatible `rexml` and `ruby` versions would be ideal.

> `scl enable rh-ruby25 bash`

I guess we could add this to [eng/docker/rhel.Dockerfile](https://github.com/dotnet/aspnetcore/blob/main/eng/docker/rhel.Dockerfile). But, wouldn't a working and modern `rpmpkg` image on `mcr` be the right way forward❔

**Comment by dougbu** [comment]
Wahoo❕

**Pull request merged**
